src/Cache/ClientCache.php

```diff
@@ -0,0 +1,48 @@
+<?php
+declare(strict_types = 1);
+
+namespace LanguageServer\Cache;
+
+use LanguageServer\LanguageClient;
+use Sabre\Event\Promise;
+
+/**
+ * Caches content through a xcache/* requests
+ */
+class ClientCache implements Cache
+{
+    /**
+     * @param LanguageClient $client
+     */
+    public function __construct(LanguageClient $client)
+    {
+        $this->client = $client;
+    }
+
+    /**
+     * Gets a value from the cache
+     *
+     * @param string $key
+     * @return Promise <mixed>
+     */
+    public function get(string $key): Promise
+    {
+        return $this->client->xcache->get($key)->then('unserialize')->otherwise(function () {
+            // Ignore
+        });
+    }
+
+    /**
+     * Sets a value in the cache
+     *
+     * @param string $key
+     * @param mixed  $value
+     * @return Promise
+     */
+    public function set(string $key, $value): Promise
+    {
+        return $this->client->xcache->set($key, serialize($value))->otherwise(function () {
```

Comment:
Re: Security. The security concerns all revolve around `unserialize` triggering autoloading and `__wakeup` calls of unwanted classes, if the serialized content has been tempered with and then relies on the presence of a class that would actually cause harm. For this, you would need to have access to the cache. If an attacker has access to the `~/.phpls`, he has write access to your user folder already, which means he could do a lot worse stuff than tampering with a PHP LS cache. And he also already has access to `~/.vscode/extensions`, which means instead of trying to inject an attack in the cache, he could just modify the PHP source in place, which would be much easier, can do a lot more harm and there is nothing we can do against it.

If, on the other hand, the cache is stored inside the client, then it is his responsibility to secure it.

> using serialize() will also mean your cache contents will be coupled to the class definitions (and future updates can make the contents inconsistent with future class definitions).

That's why all cache items are prefixed with a version. The cache format will eventually change at some point, whether I use JSON or `serialize`, it needs to be versioned.